File: playwright-pom-framework/src/test/java/com/automation/framework/utils/ExcelDataProvider.java
package com.automation.framework.utils;

import org.testng.annotations.DataProvider;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

public final class ExcelDataProvider {

  private ExcelDataProvider() {}

  private static Path excelPath() {
    String override = System.getProperty("excelPath");
    if (override != null && !override.isBlank()) {
      return Path.of(override);
    }
    return Path.of("target", "testdata", "LoginTestData.xlsx");
  }

  @DataProvider(name = "loginExcel", parallel = false)
  public static Object[][] loginExcel() {
    Path file = excelPath();
    ExcelUtils.ensureSampleLoginWorkbook(file);

    Object[][] raw = ExcelUtils.readSheet(file, "Login");
    if (raw.length == 0) {
      return raw;
    }

    List<Object[]> rows = new ArrayList<>();
    for (Object[] row : raw) {
      if (row.length < 5) {
        continue;
      }
      String run = String.valueOf(row[4]).trim();
      if ("Y".equalsIgnoreCase(run)) {
        rows.add(row);
      }
    }

    return rows.toArray(new Object[0][]);
  }
}

File: python/framework/utils/excel_data_provider.py
from pathlib import Path

from framework.utils.config_reader import get_optional
from framework.utils.excel_utils import ensure_sample_login_workbook, read_sheet
from framework.utils.paths import TARGET_DIR


def _excel_path():
    override = get_optional("excelPath")
    if override:
        return Path(override)
    return TARGET_DIR / "testdata" / "LoginTestData.xlsx"


def login_excel():
    file_path = _excel_path()
    ensure_sample_login_workbook(file_path)

    raw = read_sheet(file_path, "Login")
    if not raw:
        return []

    rows = []
    for row in raw:
        if len(row) < 5:
            continue
        run = str(row[4]).strip()
        if run.upper() == "Y":
            rows.append(tuple(row))

    return rows
